src/multipart/multipart.rs

```diff
@@ -0,0 +1,469 @@
+pub struct MultipartField {
+    name: Cow<'static, str>,
+    value: Box<Read + Send>,
+    value_length: Option<u64>,
+    mime: Option<Mime>,
+    file_name: Option<Cow<'static, str>>,
+}
+
+// TODO: MultipartField cannot derive debug because value is not Debug
+// Not sure how to best resolve this...
+impl std::fmt::Debug for MultipartField {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "")
+    }
+}
+
+impl MultipartField {
+    /// Makes a String parameter.
+    ///
+    /// ```
+    /// reqwest::MultipartField::param("key", "value");
+    /// ```
+    ///
+    /// ```
+    /// let key: String = "key".to_string();
+    /// let value: String = "value".to_string();
+    /// reqwest::MultipartField::param(key, value);
+    /// ```
+    ///
+    pub fn param<T: Into<Cow<'static, str>>, U: AsRef<[u8]> + Send + 'static>(name: T, value: U) -> MultipartField {
+        let value_length = Some(value.as_ref().len() as u64);
+        MultipartField {
+            name: name.into(),
+            value: Box::new(std::io::Cursor::new(value)),
+            value_length: value_length,
+            mime: None,
+            file_name: None,
+        }
+    }
+    /// Adds a generic reader.
+    ///
+    /// ```
+    /// use std::io::empty;
+    /// let reader = empty();
+    /// reqwest::MultipartField::reader("key", reader);
+    /// ```
+    ///
+    pub fn reader<T: Into<Cow<'static, str>>, U: Read + Send + 'static>(name: T, value: U) -> MultipartField {
+        MultipartField {
+            name: name.into(),
+            value: Box::from(value),
+            value_length: None,
+            mime: None,
+            file_name: None,
+        }
+    }
+    /// Makes a file parameter.
+    /// Defaults to mime type application/octet-stream.
+    ///
+    /// ```no_run
+    /// reqwest::MultipartField::file("key", "/path/to/file");
+    /// ```
+    ///
+    /// # Errors
+    /// Errors when the file cannot be opened.
+    pub fn file<T: Into<Cow<'static, str>>, U: AsRef<std::path::Path>>(
+        name: T,
+        path: U,
+    ) -> std::io::Result<MultipartField> {
+        let file_name = path.as_ref().file_name().and_then(|filename| {
+            Some(Cow::from(filename.to_string_lossy().into_owned()))
+        });
+        let file_length = std::fs::metadata(path.as_ref())
+            .ok()
+            .and_then(|metadata| Some(metadata.len() as u64));
+        Ok(MultipartField {
+            name: name.into(),
+            value: Box::new(std::fs::File::open(path.as_ref())?),
+            value_length: file_length,
+            mime: Some(::hyper::mime::APPLICATION_OCTET_STREAM),
+            file_name: file_name,
+        })
+    }
+    /// Sets the mime, builder style.
+    ///
+    /// ```
+    /// use reqwest::mime;
+    /// reqwest::MultipartField::param("key", "value").mime(Some(mime::IMAGE_BMP));
+    /// ```
+    ///
+    pub fn mime(mut self, mime: Option<Mime>) -> MultipartField {
```

Comment:
Is the `Option` to allow for someone to remove the mime after setting it? If there's no need to remove a mime, the `Option` could be removed...

If there is a need, this could be `mime<T: Into<Option<Mime>>` and then `mime(mime::IMAGE_BMP)` would work.